django_tutorial/first_project/web_app/static_src/js/redux/color.js

```diff
@@ -0,0 +1,59 @@
+import {getRegions} from './regions'
+import {Quantiler} from '../utilities'
+
+/*
+ * Note that the color controller uses a 'slice' of the global state, i.e. the state
+ * we deal with here is stored in the 'color' field of the global state, or in other
+ * words, stateFromThisFile === globalState.color.
+ */
+
+
+ ////////////////// Initial State //////////////////////
+
+
+const initialState = {
+  colorField: 'sba_per_small_bus',
+  colorQuantiler: undefined,
+  numQuantiles: 10
+}
+
+
+
+////////////////// State Selectors //////////////////////
+
+
+export const getColorField         = (state) => state.colorField
+export const getNumColorQuantiles  = (state) => state.numQuantiles
+export const getColorQuantiler     = (state) => state.colorQuantiler
+
+
+////////////////// Action Creators //////////////////////
+
+export function setColorField(colorField) {
+  return {type: 'SET_COLOR_FIELD', colorField}
+}
+
+
+////////////////// Reducers //////////////////////
+
+/**
+ * Note that unlike a normal reducer, this reducer also accepts an optional regionState
+ * param since the color quantiler state depends on the region state
+ */
+export default function colorReducer(state=initialState, action={}, regionState) {
+  const {type, colorField=getColorField(state)} = action
+  switch(type) {
+    case 'SET_COLOR_FIELD':
+    case 'SET_DISTRICT_REGION_TYPE_AND_REGIONS':
+      const regions = regionState===undefined ? [] : Object.values(getRegions(regionState))
+      return {
+        ...state,
+        colorField,
+        colorQuantiler: new Quantiler(
```

Comment:
I'll need you to explain what the Quantiler constructor is doing

Reply:
The Quantiler takes an array of data and calculates its [quantiles](https://en.wikipedia.org/wiki/Quantile), so the views can use quantile.getQuantile(value) to get the quantile of any data point.  For instance the map regions are each assigned one of 10 colors based on the value of that region.  There's a little more info on the Quantiler class definition in utilities.js, but I see that its description got truncated so I'll fix that in a commit shortly.

Also, the quantiler is technically derived data (it can be reproduced from the colorField and the set of regions), so if/when we add the Reselect library we can then pull the quantiler out of the state and just have its selector create the Quantiler on demand.

Reply:
Just updated the Quantiler docs.